src/misc/programs.cpp

```diff
@@ -39,43 +41,51 @@ Bitu call_program;
 
-static Bit8u exe_block[]={
-	0xbc,0x00,0x04,					//MOV SP,0x400 decrease stack size
-	0xbb,0x40,0x00,					//MOV BX,0x040 for memory resize
-	0xb4,0x4a,						//MOV AH,0x4A	Resize memory block
-	0xcd,0x21,						//INT 0x21
-//pos 12 is callback number
-	0xFE,0x38,0x00,0x00,			//CALLBack number
-	0xb8,0x00,0x4c,					//Mov ax,4c00
-	0xcd,0x21,						//INT 0x21
+constexpr uint8_t exe_block[] = {
+        0xbc, 0x00, 0x04,       // MOV SP,0x400  Decrease stack size
+        0xbb, 0x40, 0x00,       // MOV BX,0x040  For memory resize
+        0xb4, 0x4a,             // MOV AH,0x4A   Resize memory block
+        0xcd, 0x21,             // INT 0x21
+        0xFE, 0x38, 0x00, 0x00, // 12th position is the CALLBack number
+        0xb8, 0x00, 0x4c,       // Mov ax,4c00
+        0xcd, 0x21,             // INT 0x21
 };
 
-#define CB_POS 12
+// COM data constants
+constexpr int callback_pos = 12;
+constexpr int com_data_size = 32;
 
-static std::vector<PROGRAMS_Main*> internal_progs;
+// Persistent program containers
+using com_data_t = std::unique_ptr<uint8_t[]>;
+static std::vector<com_data_t> internal_progs_comdata;
+static std::vector<PROGRAMS_Main *> internal_progs;
 
-void PROGRAMS_MakeFile(char const * const name,PROGRAMS_Main * main) {
-	Bit8u * comdata=(Bit8u *)malloc(32); //MEM LEAK
+void PROGRAMS_MakeFile(const char *name, PROGRAMS_Main *main)
+{
+	com_data_t comdata = std::make_unique<uint8_t[]>(com_data_size);
+	assert(comdata);
 
-	if (comdata == nullptr) {
-		E_Exit("Could not allocate memory for com-file data.");
-	}
+	memcpy(comdata.get(), &exe_block, sizeof(exe_block));
+	comdata[callback_pos] = static_cast<uint8_t>(call_program & 0xff);
+	comdata[callback_pos + 1] = static_cast<uint8_t>((call_program >> 8) & 0xff);
 
-	memcpy(comdata,&exe_block,sizeof(exe_block));
-	comdata[CB_POS]=(Bit8u)(call_program&0xff);
-	comdata[CB_POS+1]=(Bit8u)((call_program>>8)&0xff);
+	// Save the current program's vector index in its COM data
+	assert(internal_progs.size() <= UINT8_MAX);
+	const auto index = static_cast<uint8_t>(internal_progs.size());
+	comdata[sizeof(exe_block)] = index;
 
-	/* Copy save the pointer in the vector and save it's index */
-	if (internal_progs.size() > 255) E_Exit("PROGRAMS_MakeFile program size too large (%d)",static_cast<int>(internal_progs.size()));
-	Bit8u index = (Bit8u)internal_progs.size();
-	internal_progs.push_back(main);
-
-	memcpy(&comdata[sizeof(exe_block)],&index,sizeof(index));
-	Bit32u size=sizeof(exe_block)+sizeof(index);	
-	VFILE_Register(name,comdata,size);	
-}
+	// Register the COM program with the Z:\ virtual filesystem
+	const auto file_size = static_cast<uint32_t>(sizeof(exe_block) +
+	                                             sizeof(index));
+	VFILE_Register(name, comdata.get(), file_size);
 
+	// Register the COM data's unique_ptr to prevent it from going out of scope
+	internal_progs_comdata.push_back(std::move(comdata));
 
```

Comment:
Our `comdata` smart-pointer will free it's memory as soon as it goes out of scope - which normally would be the end of this function.  So instead, we move it into the `internal_progs_comdata` global vector to prevent `comdata` from going out of scope.  Eventually when the entire program is shutdown, the vector then destroys all of its elements and then frees its own space.